File: cuda_bindings/benchmarks/benchmarks/bench_ctx_device.py
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
#
# SPDX-License-Identifier: Apache-2.0

import time

from runner.runtime import ensure_context

from cuda.bindings import driver as cuda

CTX = ensure_context()

_, DEVICE = cuda.cuDeviceGet(0)
ATTRIBUTE = cuda.CUdevice_attribute.CU_DEVICE_ATTRIBUTE_COMPUTE_CAPABILITY_MAJOR


def bench_ctx_get_current(loops: int) -> float:
    _cuCtxGetCurrent = cuda.cuCtxGetCurrent

    t0 = time.perf_counter()
    for _ in range(loops):
        _cuCtxGetCurrent()
    return time.perf_counter() - t0


def bench_ctx_set_current(loops: int) -> float:
    _cuCtxSetCurrent = cuda.cuCtxSetCurrent
    _ctx = CTX

    t0 = time.perf_counter()
    for _ in range(loops):
        _cuCtxSetCurrent(_ctx)
    return time.perf_counter() - t0


def bench_ctx_get_device(loops: int) -> float:
    _cuCtxGetDevice = cuda.cuCtxGetDevice

    t0 = time.perf_counter()
    for _ in range(loops):
        _cuCtxGetDevice()
    return time.perf_counter() - t0


def bench_device_get(loops: int) -> float:
    _cuDeviceGet = cuda.cuDeviceGet

    t0 = time.perf_counter()
    for _ in range(loops):
        _cuDeviceGet(0)
    return time.perf_counter() - t0


def bench_device_get_attribute(loops: int) -> float:
    _cuDeviceGetAttribute = cuda.cuDeviceGetAttribute
    _attr = ATTRIBUTE
    _dev = DEVICE

    t0 = time.perf_counter()
    for _ in range(loops):
        _cuDeviceGetAttribute(_attr, _dev)
    return time.perf_counter() - t0

File: cuda_bindings/benchmarks/benchmarks/cpp/bench_ctx_device.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
//
// SPDX-License-Identifier: Apache-2.0

#include <cuda.h>

#include "bench_support.hpp"

#include <cstdlib>
#include <iostream>


static void check_cu(CUresult status, const char* message) {
    if (status != CUDA_SUCCESS) {
        const char* error_name = nullptr;
        cuGetErrorName(status, &error_name);
        std::cerr << message << ": " << (error_name ? error_name : "unknown") << '\n';
        std::exit(1);
    }
}


int main(int argc, char** argv) {
    bench::Options options = bench::parse_args(argc, argv);
    if (options.benchmark_name.empty()) {
        options.benchmark_name = "cpp.ctx_device.ctx_get_current";
    }

    // Setup: init CUDA and create a context
    check_cu(cuInit(0), "cuInit failed");

    CUdevice device;
    check_cu(cuDeviceGet(&device, 0), "cuDeviceGet failed");

    CUcontext ctx;
    CUctxCreateParams ctxParams = {};
    check_cu(cuCtxCreate(&ctx, &ctxParams, 0, device), "cuCtxCreate failed");

    CUcontext current_ctx = nullptr;

    // Run benchmark
    auto results = bench::run_benchmark(options, [&]() {
        check_cu(
            cuCtxGetCurrent(&current_ctx),
            "cuCtxGetCurrent failed"
        );
    });

    // Sanity check: the call actually returned our context
    if (current_ctx != ctx) {
        std::cerr << "unexpected: cuCtxGetCurrent returned a different context\n";
    }

    // Cleanup
    check_cu(cuCtxDestroy(ctx), "cuCtxDestroy failed");

    // Output
    bench::print_summary(options.benchmark_name, results);

    if (!options.output_path.empty()) {
        bench::write_pyperf_json(options.output_path, options.benchmark_name, options.loops, results);
    }

    return 0;
}
